### ISSUE TITLE
Our Diff class sucks --> Switch to "modified-file coordinate system"

### ISSUE THREAD

**Issue by Makman2** [coala/coala#3885]
When changing lines inside a `Diff` object, it uses the old lines as reference. It also doesn't allow a modification to a line that was once modified. This behaviour highly sucks and should be changed to pick up the "new" lines and their respective line numbers, and modifications would work on these.

I already had problems with that, and currently @satwikkansal is implementing a `from_unified_diff` method that generates a `Diff` object out of a unified diff patch. I already wondered why his code is so damn complicated and tried to implement a shorter prototype, and though I think his code is even too complicated for our current `Diff`, complexity could be way more reduced by using the "modified-file coordinate system".

\+ imo this behaviour would be way more intuitive and also allows easily multiple additions and deletions.

Please let's make our diff not sucking any more...

**Comment by arjunsinghy96** [comment]
I would like to work on this.

**Comment by Makman2** [comment]
Ah wait a sec, I wanted to discuss that with @fneu as he's the author :+1: :3

**Comment by Makman2** [comment]
please wait with implementation @arjunsinghy96

**Comment by meetmangukiya** [comment]
> Our Diff class sucks 

lol (Subscribing)

**Comment by fneu** [comment]
I'm not the author :D

That would be @sils. I agree with many of the points here made so far.

My only concern is that things like the original line being saved might be used when 

- determining the compatibility of two diffs (merge conflicts?)
- determining which coala results are new (files may have changed)
- possibly other, similar functionality.

That said, I don't see a reason why we couldn't use a more standard diff approach, but expect a considerable amount of work :/

**Comment by fneu** [comment]
> It also doesn't allow a modification to a line that was once modified.

Interesting point, I've never seen it this way. Adding a second diff with the desired change should work as long as the changes are mergeable on the specific line?

2 modifications to the same line are a classic merge conflict that we can usually resolve, though, because we actually compare the line content and update the change IIRC.

**Comment by sils** [comment]
Yep of line conflicts occur essentially we convert both lines into lists of one char strings and make another diff out of this, merge it like usual which is pretty powerful already

**Comment by Makman2** [comment]
Okay seems people are agreeing, let's do it :+1:

**Comment by Makman2** [comment]
@arjunsinghy96 you can go ahead :)

**Comment by Makman2** [comment]
and sorry @fneu seems I remembered wrongly ... happens :3

**Comment by arjunsinghy96** [comment]
I will go ahead and try to solve this then.

**Comment by arjunsinghy96** [comment]
Do we need to merge both the modifications or just keep the latest modification.

**Comment by Makman2** [comment]
if there's an unresolvable conflict, throw an error. Though we support merging stuff inside a single line under certain conditions.

**Comment by arjunsinghy96** [comment]
Yeah. When we try to modify same line again, our Diff converts both replacement lines into list of char, create Diff object for both and try to merge them(or add). I found that almost every time we try to merge two Diffs of char we get a conflict.

**Comment by arjunsinghy96** [comment]
What if we apply modification on the last modified line.

**Comment by Makman2** [comment]
it refers again to that line, that's the principle of the modified-file-coordinate-system :)

**Comment by Makman2** [comment]
hm yeah that's the challenge... you'll find a way :)

**Comment by arjunsinghy96** [comment]
Can you point me to some reference for `modified line coordinate system` . I need more insight into it.

**Comment by Makman2** [comment]
What I mean is basically this:

Currently diffs use - I call it - "original file coordinate system". That means any change is performed relative to the original file. Imagine this:
```
This is the original file
abcdef
```
and I add a line:
```
This is the original file
123
abcdef
```
If we want to add another line of text, regardless whether we do this before `123` or after, current `Diff` throws an error, as it doesn't know where to place the new line respective to `123` (we just can choose line 0, 1 and 2 from the original file), as every change refers to the original file.

The "modified-file coordinate system" uses the modified version of the file the `Diff` currently maintains. Every change happens in the context of every change you've already made.

Example from before: We add `123` again, and then we try to add `xyz` after `123`. This is totally possible, we refer to the new line now: 2. Then we want to add something after `xyz`. No problem, `xyz` has become line 3, so we tell `Diff` to add something after line 3.

GitHub displays also on the left side the old lines and right beneath the new lines, when you review commits :+1:

**Comment by arjunsinghy96** [comment]
Okay. Thanks, now I know I was moving in right direction. Time to think how to implement this. (Maybe by keeping a modified version in the `Diff`) Right now we only create the modified file when `modified` method is called.

**Comment by satwikkansal** [comment]
@arjunsinghy96  I think this involves maintaining an `original_file_modified_till_now` version of file in the `Diff` class and instead of using `original_file` as reference when doing modifications (like `add_line()`, `modify_line()`), use this `original_file_modified_till_now` as reference. 
This `original_file_with_modifications_till_now` will be updated for every addition or deletion to the `Diff` object.

**Comment by arjunsinghy96** [comment]
@satwikkansal This is what I was thinking. But there is a problem.
Suppose our original file was
```
this is original
second line
```
now if we add `inserted` after `this is original`. Our `modified_till_now` will be
```
this is original
insertion
second line
```
Now if a instruction to modify_line 2 is given, which line( `insertion` or `second line`) it is refering.
Then the code using `Diff` will have to keep track of changes.

**Comment by sils** [comment]
@Makman2 you wanna introduce diffs based on already modified files? That won't work, the bears run in parallel and the patches could be applied in any order by the user, you can't just allow them sequentially.

Frankly I don't see what this issue is wanting to achieve, our diffs are pretty smart, maybe it can be improved a little but I don't see any cases where I could objectively make better decisions than that algorithm we have off the top of my head.

**Comment by Makman2** [comment]
1. There's already `Diff.modified` which contains the modified version.
2. > Now if a instruction to modify_line 2 is given, which line( insertion or second line) it is refering.
   > Then the code using Diff will have to keep track of changes.
   
   like said, it refers to `insertion` as `insertion` has become line 2.
3. @sils then it's the challenge to merge them like we do now. This is not really different from before, just the data representation is different. It should be possible to construct changes relative to the original file.

> Frankly I don't see what this issue is wanting to achieve, our diffs are pretty smart, maybe it can be improved a little but I don't see any cases where I could objectively make better decisions than that algorithm we have off the top of my head.

Diffs ain't smart regarding usability. It's common to add changes onto a line that was already changed. It's not about changing the behaviour of coala's patch applying

**Comment by bhavishyagopesh** [comment]
@Makman2 @sils I would like to take  this up?I understand the issue so what do you say?

**Comment by Makman2** [comment]
It's currently assigned, ask @arjunsinghy96 :+1:

**Comment by bhavishyagopesh** [comment]
@arjunsinghy96 hey I'm eyeing for G-SoC(deadline's close too), So if you can please allow I would like to work on this.

**Comment by Makman2** [comment]
@bhavishyagopesh though seems you are new here, you sure you want to start with a medium issue? That's really not that easy^^

**Comment by bhavishyagopesh** [comment]
@Makman2 I would like to give it a try and I have got my commits merged to sympy,hyper,mailman(which are python based), so maybe I could do something.Or if you could direct (that would be huge help)me to issues which  are much more appropriate with G-SoC in mind.Thank You.

**Comment by bhavishyagopesh** [comment]
@Makman2 So what do you say?

**Comment by arjunsinghy96** [comment]
@bhavishyagopesh I am currently working on this. You could help me if you like. I am also a GSoC aspirant.

**Comment by satwikkansal** [comment]
@arjunsinghy96 Still on this?

**Comment by arjunsinghy96** [comment]
@satwikkansal Yes, I am still working with this. I was busy with GSoC Proposal. Also, I think simply adding a `modified_till_now` won't be that effective.

**Comment by Makman2** [comment]
> Also, I think simply adding a modified_till_now won't be that effective.

**Comment by meetmangukiya** [comment]
Is the discussion to keep this issue resolved? If not, please add `needs design` label :P

**Comment by meetmangukiya** [comment]
)

**Comment by Makman2** [comment]
@meetmangukiya I think the discussion is about some implementation details, but imo the task is clear in this state :) So no `needs-design` label needed I believe ;)

**Comment by sils** [comment]
> Diffs ain't smart regarding usability. It's common to add changes onto a line that was already changed. It's not about changing the behaviour of coala's patch applying

It definitely doesn't make sense to change the way the diffs behave. What could be done would be methods that allow modifying a diff based on the modified version in addition to what we have, those would just be wrapper methods around existing functionality (not that I see a usecase for it).

This has nothing to do with coala applying patches and can't affect bears that yield multiple diffs because they have to be independent from each other anyway.

**Comment by Vedarth** [comment]
@Makman2 I would like to take up this issue. Related to a issue which I am already working on :wink:

**Comment by Makman2** [comment]
> (not that I see a usecase for it).

I do, I tried several times writing bears where it was a pain to aggregate changes between two lines before applying them to the diff. (sorry for the late response, I missed it :/)

About wrapping current functions: Yes maybe we have to do this for the sake of compatibility. Can be designed out properly though (maybe by writing a new version of the diff-class newer bears can use?).

**Comment by Makman2** [comment]
@Vedarth please properly design it, not so easy to integrate into our ecosystem :+1:

**Comment by Vedarth** [comment]
@Makman2 So I will need to write a new version of `diff` class, according to https://github.com/coala/coala/issues/3885#issuecomment-372126115 From my understanding we need to make `diff` remember the changes i.e. use `Diff.modified`, which it already has. So it means I will just need to supply `modified diff` instead of original file in all the methods? That is all I think I need to do. Am I thinking correctly or something more needs to be done?

**Comment by Makman2** [comment]
No that's not what the issue is about. Please read it again^^

**Comment by Vedarth** [comment]
> The "modified-file coordinate system" uses the modified version of the file the Diff currently maintains. Every change happens in the context of every change you've already made.

I got that from this. Anyways, in plain and simple language, we need to make diff refer to the latest modified file instead of original file when making changes. But we will also need to handle conflicts when doing that.
Also in https://github.com/coala/coala/issues/3885#issuecomment-286068295 you said insertion will become line 2 that means we will need the modified file. How else can I make diff remember the changes. I understand the main issue is that a line which was already modified cannot be modified again which becomes quite tiresome. So the solution that I thought was using the modified diff instead of original while doing modifications. If you think this approach is incorrect kindly suggest changes.

**Comment by Makman2** [comment]
>  So the solution that I thought was using the modified diff instead of original while doing modifications

Yes correct :)

> No that's not what the issue is about. Please read it again^^

Sorry not explaining further, I was a bit tired then :3 I probably haven't understood you completely too^^

**Comment by Vedarth** [comment]
@Makman2 I will submit a pr :smile:
